crypto/hpke/hpke.c

```diff
@@ -0,0 +1,3137 @@
+    *p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_SALT,
+                                             (unsigned char *)salt, saltlen);
+    *p = OSSL_PARAM_construct_end();
+    if (EVP_KDF_CTX_set_params(kctx, params) <= 0) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    lsecretlen = EVP_KDF_CTX_get_kdf_size(kctx);
+    if (lsecretlen > *secretlen) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /* Do the derivation */
+    if (EVP_KDF_derive(kctx, secret, lsecretlen, params) <= 0) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    *secretlen = lsecretlen;
+
+err:
+    EVP_KDF_free(kdf);
+    EVP_KDF_CTX_free(kctx);
+    memset(labeled_ikmbuf, 0, OSSL_HPKE_MAXSIZE);
+    return erv;
+}
+
+/*
+ * @brief RFC5869 HKDF-Expand
+ *
+ * @param libctx is the context to use (normally NULL)
+ * @param suite is the ciphersuite
+ * @param mode5869 - controls labelling specifics
+ * @param prk - the initial pseudo-random key material
+ * @param prk - length of above
+ * @param label - label to prepend to info
+ * @param labellen - label to prepend to info
+ * @param context - the info
+ * @param contextlen - length of above
+ * @param L - the length of the output desired
+ * @param out - the result of expansion (allocated by caller)
+ * @param outlen - buf size on input
+ * @return 1 for good otherwise bad
+ */
+static int hpke_expand(OSSL_LIB_CTX *libctx,
+                       const OSSL_HPKE_SUITE suite, const int mode5869,
+                       const unsigned char *prk, const size_t prklen,
+                       const char *label, const size_t labellen,
+                       const unsigned char *info, const size_t infolen,
+                       const uint32_t L,
+                       unsigned char *out, size_t *outlen)
+{
+    int erv = 1;
+    unsigned char libuf[INT_MAXSIZE];
+    unsigned char *lip = libuf;
+    size_t concat_offset = 0;
+    size_t loutlen = L;
+    EVP_KDF *kdf = NULL;
+    EVP_KDF_CTX *kctx = NULL;
+    OSSL_PARAM params[5], *p = params;
+    int mode = EVP_PKEY_HKDEF_MODE_EXPAND_ONLY;
+    const char *mdname = NULL;
+    uint16_t kem_ind = 0;
+    uint16_t kdf_ind = 0;
+
+    if (L > *outlen) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /* Handle oddities of HPKE labels (or not) */
+    switch (mode5869) {
+    case OSSL_HPKE_5869_MODE_PURE:
+        if ((labellen + infolen) >= INT_MAXSIZE) {
+            OSSL_HPKE_err;
+            goto err;
+        }
+        memcpy(lip, label, labellen);
+        memcpy(lip + labellen, info, infolen);
+        concat_offset = labellen + infolen;
+        break;
+
+    case OSSL_HPKE_5869_MODE_KEM:
+        lip[0] = (L / 256) % 256;
+        lip[1] = L % 256;
+        concat_offset = 2;
+        memcpy(lip + concat_offset, OSSL_HPKE_VERLABEL,
+               strlen(OSSL_HPKE_VERLABEL));
+        concat_offset += strlen(OSSL_HPKE_VERLABEL);
+        if (concat_offset >= INT_MAXSIZE) {
```

Comment:
Check is too late. Same comment applies below.

Reply:
yeah fair enough, will fix or move to WPACKET